Robust locking mechanism

broskoTT opened the pull request:
### Issue
Fixes #691 #654 #651, also completes #72 #523 , related to #418

### Description
We've had this problem for a while: Process crashing could leave locked mutexes on the system. The workaround was to do "rm /dev/shm/*".
After some refactoring and an attempt to consolidate all locking code to a single class, the non proper cleanup has hit us yet again. I've finally dedicated time to try to solve this properly.
Of course some of the usages of lock might disappear by the future redesign but we still might need some interprocess critical sections in UMD, and this class should be able to provide it robustly.
The solution is to use PTHREAD_MUTEX_ROBUST attribute, which allows for a mechanism to recover when a process which holds the lock crashes. This also meant that we had to manually write locking code, which made this effort  more involving.

### List of the changes
- Implemented RobustLock class. It is intended to be used as RAII.
- LockManager is now just a simple wrapper which concatenates strings.
- Removed all initialize and clear_mutex calls, as this is now not needed. Each "locking" will also initialize the mutex. This part might move to LockManager if we deem this is not performant enough.
- Removed initialize_default_chip_mutexes since its not needed anymore.
- Renamed get_mutex to acquire_lock to better communicate the API
- Removed includes and usages to boost::interprocess.
- Re-enabled multiprocess tests

### Testing
I've introduced crashes at many places in the code, and added some prints to all codepaths in RobustMutex. I was then manually running some test and observed what was happening. Verified everything worked as intended, verified no deadlocks occurred, verified mutex was successfully restored after crash. Also, I reenabled multiprocess tests which were previously consistently failing with the old locks, and verified they pass now, so it seems like the locks are indeed locking critical sections.

### API Changes
There are no API changes in this PR.
But removal of #include <boost/interprocess/sync/named_mutex.hpp> will likely make some people happy.

pjanevskiTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,149 @@
+    }
+    err = pthread_mutex_init(mutex_ptr, &attr);
+    if (err != 0) {
+        throw std::runtime_error("pthread_mutex_init failed: " + std::to_string(err));
```
might be useful to store mutex name as well so we can now which one failed. Maybe that information can be useful sometimes for debugging if needed

pjanevskiTT commented on device/api/umd/device/utils/robust_mutex.h:
```diff
@@ -0,0 +1,54 @@
+private:
+    // Does everything related to initializing the mutex, even on first time creation.
+    void initialize_pthread_mutex(std::string mutex_name);
+    // Closes the mutex, doesn't remove the backing mutex file.
```
nit: space between previous function and comment, for all functions

pjanevskiTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,149 @@
+#include <stdexcept>
+
+static constexpr int ALL_RW_PERMISSION = 0666;
+static constexpr std::string_view UMD_LOCK_PREFIX = "umdlock.";
```
nit: add TT prefix. Not an official rule of course, maybe we can start adopting it

broskoTT replied:
Why? Do you want me to put it in tt::umd:: namespace instead?

pjanevskiTT replied:
@tt-vjovanovic asked me to put it in mutex names some time ago, not sure if we should do it as I said. I am fine with keeping this as it is

pjanevskiTT reviewed (approved):
amazing 🚀

joelsmithTT commented on device/api/umd/device/utils/lock_manager.h:
```diff
@@ -30,37 +26,40 @@ enum class MutexType {
     void initialize_mutex(MutexType mutex_type, const bool clear_mutex);
     void clear_mutex(MutexType mutex_type);
-    std::unique_lock<boost::interprocess::named_mutex> get_mutex(MutexType mutex_type);
+    std::unique_ptr<RAIIMutex> acquire_mutex(MutexType mutex_type);
```
Idiomatic C++ generally distinguishes between the mutex (the thing that gets locked/unlocked) and the lock (the wrapper that provides exception-safe lock/unlock semantics).  Please consider:
1. Making `RobustMutex` Lockable (essentially provide public lock() and unlock() methods - perhaps by renaming lock_mutex/unlock_mutex).
2. Replacing `RAIIMutex` with standard guards: `RAIIMutex` duplicates RAII pattern provided by `std::lock_guard` and `std::unique_lock`.  If `RobustMutex` were Lockable (per point 1), `RAIIMutex` becomes redundant and can be removed.
3. Instead of `acquire_mutex` returning a `std::unique_ptr<RAIIMutex>` (which involves a heap allocation), it could return a standard RAII lock guard _by value_.

broskoTT replied:
I ended up returning a unique_lock, but that one should also involve a heap allocation. lock_guard can't be used when you want to return it from a function.

joelsmithTT commented on device/api/umd/device/utils/lock_manager.h:
```diff
@@ -30,37 +26,40 @@ enum class MutexType {
     CREATE_ETH_MAP,
 };
 
+// Note that the returned RAIIMutex should never outlive the LockManager which holds underlying RobustMutexes.
```
The lifetime documentation here is great.

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,149 @@
+
+using namespace tt::umd;
+
+RobustMutex::RobustMutex(std::string mutex_name) { initialize_pthread_mutex(mutex_name); }
```
There are three options to avoid copying memory:
1. Use `const std::string&` and do copy only when you need to store that value
2. Use `std::string` and then use `std::move` when you don't need it (in your example `initialize_pthread_mutex(std::move(mutex_name));` )
3. Use `std::string_view` and do copy only when you need to store that value...

broskoTT replied:
Changed to string_view

joelsmithTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,159 @@
+void RobustMutex::initialize_pthread_mutex(std::string mutex_name) {
+    bool created = open_shm_file(mutex_name);
+    // There is a slight chance of race condition here, where first process creates the file, and second process opens
+    // it. Then the second process "surpasses" the first process and tries using the mutex before it is initialized.
```
So:
* Process A creates the shared memory file
* Process B opens it, maps it, and tries to use the `pthread_mutex_t`
* Process A finishes calling `initialize_pthread_mutex` via `initialize_pthread_mutex_first_use`
This seems like a critical flaw.

broskoTT replied:
Yes. But it seems unlikely enough for me to be fine with it for now. Note that also that happens only the first time on the system (after boot), not on each execution

joelsmithTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,159 @@
+    if (shm_fd == -1 && errno == ENOENT) {
+        // There is a slight chance of race condition here, since two processes can enter this same codepath, and then
+        // both can try creating the file.
+        shm_fd = shm_open(shm_file_name.c_str(), O_RDWR | O_CREAT, ALL_RW_PERMISSION);
```
`man 3 shm_open` says,

"O_EXCL If  O_CREAT was also specified, and a shared memory object with the given name already exists, return an error.  The check for the existence of the object, and its creation if it does not exist, are performed atomically."

Can this do the open atomically and handle EEXIST?

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,149 @@
+bool RobustMutex::open_shm_file(std::string mutex_name) {
+    std::string shm_file_name = std::string(UMD_LOCK_PREFIX) + mutex_name;
+    bool created = false;
+    shm_fd = shm_open(shm_file_name.c_str(), O_RDWR, ALL_RW_PERMISSION);
```
You should probably switch to this order:
```cpp
shm_fd = shm_open(..., O_RDWR | O_CREAT | O_EXCL, ...);
if (shm_fd == -1 && errno == EEXIST) {
    shm_fd = shm_open(..., O_RDWR, ...);
}
```
To minimize race condition in file creation.

broskoTT replied:
Yes, did exactly this and added comment to explain.

joelsmithTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,159 @@
+    // Create a pthread_mutex based on the shared memory file descriptor
+    void *addr = mmap(NULL, sizeof(pthread_mutex_t), PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
+    if (addr == MAP_FAILED) {
+        throw std::runtime_error("mmap failed for mutex " + mutex_name_ + " errno: " + std::to_string(errno));
```
Let's assume `mmap` fails.  The exception propagates out of `open_pthread_mutex`, then out of `initialize_pthread_mutex` and finally out of `RobustMutex::RobustMutex`.  Because construction failed, the object is never fully-formed.  Therefore `RobustMutex::~RobustMutex` is never executed.  Since destructor is never executed, `close_mutex()` is never called.  So the process still holds the open fd referring to the shared memory object, but the int holding the fd is gone.

I think there are some other resource leaks on construction failure paths.

broskoTT replied:
The process closes file descriptor automatically when it crashes (or rather, the os closes it when the process crashes), right?

joelsmithTT replied:
That's right, the kernel will clean up the resources no matter what.  The resource leak concern is applicable if the process continues running.  For example, it is conceivable that the higher-level application could catch the exception and try again.

joelsmithTT reviewed (commented):
Resolving the identified race conditions during initialization is crucial for ensuring the mutex's reliability under concurrency.

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,149 @@
+    bool created = open_shm_file(mutex_name);
+    // There is a slight chance of race condition here, where first process creates the file, and second process opens
+    // it. Then the second process "surpasses" the first process and tries using the mutex before it is initialized.
+    if (created) {
```
You can have two threads/processes, one will create file, one will open file. (even after comment I left in `open_shm_file`) You will have no guarantees who will get here first (one that opened file or one that creates it, it depends on CPU scheduler). So, what you need to do after opening is to lock the file, check if it is empty (that means it is just created) and do initialization.

broskoTT replied:
Hopefully did this right.

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,149 @@
+    // it. Then the second process "surpasses" the first process and tries using the mutex before it is initialized.
+    if (created) {
+        // File needs to be resized first time it is created.
+        if (ftruncate(shm_fd, sizeof(pthread_mutex_t)) != 0) {
```
Is `pthread_mutex_t` something that can be changed during OS runtime? Like you upgrade something with `sudo apt upgrade` and you get different size of it? Or even worse, you get structure of the same size, but different layout?
If that can happen, you should consider versioning mutex files. Like you have `umdlock.` as prefix, consider adding `pthread.h` "version" in the mutex name as well...

broskoTT replied:
this would never had occurred to me in three lifetimes

broskoTT replied:
You know what, if I do what you suggested, I fear that it may happen I have two processes running on different pthread versions (one long running still running on old version). Instead, I'll guard this behavior by just reporting if the size is different, since I have to check the size each time when I open the file anyway. Let me know what you think of this

broskoTT replied:
I added a comment to the class docs that this might happen, and we don't handle it well.

broskoTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,149 @@
+
+RobustMutex::RobustMutex(std::string mutex_name) { initialize_pthread_mutex(mutex_name); }
+
+RobustMutex::~RobustMutex() { close_mutex(); }
```
You'd suggest warnings in that case?

broskoTT replied:
Changed to warnings and marked noexcept

broskoTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,159 @@
+
+RAIIMutex::RAIIMutex(RobustMutex *mutex) : mutex_ptr(mutex) { mutex_ptr->lock_mutex(); }
+
+RAIIMutex::~RAIIMutex() { mutex_ptr->unlock_mutex(); }
```
Removed RAIIMutex and used unique_lock instead

broskoTT commented:
@joelsmithTT @tt-vjovanovic thanks for the thorough review!
I've addressed the comments, and the logic has changed now a bit, please take another look now, should be better.

Here's the comment on the change:
I've removed the race condition on file creation by inverting the order of trying to create a file and trying to open a file.
I've created a critical section for pthread initialisation, and with that removed the race condition where first process creates the file, but then another surpasses it and tries using pthread before it is initialized. The critical section was done using flock and static pthread_mutex_t
Changed throws into log_fatal
Changed throws in destructor into warnings, marked it noexcept to be explicit we don't want to throw on destruction path.
Renamed functions in RobustMutex so it implements BasicLockable requirement
Remove RAIIMutex, in favour of unique_lock which is in standard library
Map in LockManager now holds references to objects, rather than unique_ptrs. I had to implement move constructor and assignment operator in RobustMutex to allow that.
Changed arg mutex_name to string_view

joelsmithTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,205 @@
+
+    // Verify file size.
+    if (file_size != sizeof(pthread_mutex_t)) {
+        log_fatal(
```
`log_fatal` is broken.  It is not fatal, it just throws a `std::runtime_error`.  It's conceivable that an application could catch the exception and continue running.  This is not a problem with your code here, just pointing it out for awareness.  Something that is named `log_fatal` should actually be fatal.

I think it is appropriate to treat errors in this class as fatal (i.e. the program simply terminates).

broskoTT replied:
But the user of the lib might want to actually wrap everything in a try catch block, and do something else if UMD initialization fails. That should be why the exception mechanism was added, to add more choice on how to handle failures, compared to just having exit(1) or something like that.

joelsmithTT replied:
Fatal means causing death.  `log_fatal` should be calling `std::terminate()`, but it just throws.  The program can escape death as long as there is a try block that can catch a `std::runtime_error`.  `log_fatal` is a lie.

It's fine if RobustMutex throws exceptions and leaves the choice up to the caller.  In that case:
* `log_fatal` is a confusing way to convey this behavior
* The code would need to be made exception safe -- for example, if `mmap` fails, currently the `multithread_mutex_` doesn't get unlocked and the `shm_fd` is leaked

broskoTT replied:
Sure, log_fatal doesn't have a proper name https://github.com/tenstorrent/tt-umd/issues/696
Ok I'll try to unlock and cleanup before exception where needed...

joelsmithTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,205 @@
+    // multithread_mutex_ for all different RobustMutex instances which can affect perf of these operations, but that is
+    // fine since this is executed rarely, only on initialization and only once after booting the system. Regarding
+    // flock perf, this happens only when initializing the mutex, so it is not a big deal.
+    flock(shm_fd, LOCK_EX);
+    pthread_mutex_lock(&multithread_mutex_);
```
Should the return values of these function calls be checked?

broskoTT replied:
right, forgot about it

joelsmithTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,205 @@
+void RobustMutex::initialize_pthread_mutex(std::string_view mutex_name) {
+    bool created = open_shm_file(mutex_name);
+
+    // We need a critical section here in which we test if the mutex has been initialized, and if not initialize it.
+    // If we don't create a critical section for this, then two processes could race, one to initialize the mutex and
+    // the other one to use it before it is initialized. flock ensures only multiprocess locking, but does not guarantee
+    // multithread locking. Due to that, we need to use multithread_mutex_ which guarantees multithread locking but not
+    // multiprocess locking. Note that flock is released automatically on process crash, and static multithread_mutex_
+    // is not persistent, so we're safe even if the process crashes in the critical section. We use only a single
+    // multithread_mutex_ for all different RobustMutex instances which can affect perf of these operations, but that is
+    // fine since this is executed rarely, only on initialization and only once after booting the system. Regarding
+    // flock perf, this happens only when initializing the mutex, so it is not a big deal.
```
👍

joelsmithTT reviewed (commented):
I appreciate the attention to correctness around initialization, this looks like an effective technique.

A few housekeeping notes:

* There's a comment in `LocalChip::initialize_default_chip_mutexes` that won't be accurate after this change goes in
* The `clean_system_resources` flag (also spelled `clear_mutex`) shows up in multiple files.  With the improved robustness in mutex handling, are these flags still needed?
* Similarly, regarding the methods that delete a mutex (i.e. `LockManager::clear_mutex`, `LockManager::clear_mutex_internal`).  Is this operation still needed, after UMD does not have to worry about tidying up stale state in `/dev/shm/`?

broskoTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,205 @@
+    flock(shm_fd, LOCK_EX);
+    pthread_mutex_lock(&multithread_mutex_);
+
+    size_t file_size = get_file_size(shm_fd);
+    // This means the file was just created but nor file nor mutex were initialized.
+    bool should_initialize = (file_size == 0);
```
Sure that sounds like a breaking scenario. Any ideas how to fix it? I'd say this is unlikely to happen, so we can maybe ignore it. It is also possible that somebody manually manipulates the file in shm, which can also lead to a failure...

joelsmithTT replied:
How about:

* In the shared memory segment, store a `struct { pthread_mutex_t mutex; uint64_t magic; }` instead of a `pthread_mutex_t`
* Write a hard-coded constant random number to the magic field _after_ `pthread_mutex_init()` succeeds
* should_initialize logic checks for the presence of the magic number instead of the segment size

You are correct about external manipulation of the data.  That seems like a fragility inherent to using shared memory.  But that fragility is something we have today already, with the boost named_mutex scheme.  Your robust mutex represents a significant improvement.

broskoTT replied:
I did exactly this. I tested with some example code failing midway in initialization, both with throwing and pure exit, everything works as expected.
User gets a warning in case of corrupted underlying file. After manually deleting it, everything starts working again (it gets reinitialized)

broskoTT commented:
Changes after second iteration of addressing comments
- Added check for flock and pthread_mutex_lock during initialization.
- Also wrapped this in a small RAII class, which will guarantee cleanup of these resources even if initialization code throws.
- Moved initialize() out of constructor. This allows writing proper cleanup code in the destructor, and than knowing it will be called if initialize() fails (which is not true during constructor's execution).
- Changed simple pthread_mutex to a wrapper which also holds a magic number. The shared memory file is now mapped into this struct instead of just pthread_mutex. The struct has a bool in addition, which designates whether the pthread_mutex was properly initialized, since pthread_mutex doesn't offer any other mechanism to verify itself if it is initialized.
- Add mutex_name_ to move constuructors, this was a bug, but mutex_name was used instead of mutex_name_ so it was working, now it doesn't anymore once I separated initialize and construction.

Retested again with some cases, and running multiprocess tests.

Thanks @joelsmithTT and @tt-vjovanovic for continued support on this!

broskoTT commented:
Time taken for 10000000 mutex initializations: 40752193209 nanoseconds, so single one is around 40752
Time taken for 10000000 lock/unlock: 654239360 nanoseconds, so single one is around 654
Time taken for 10000000 unique_lock/unlock: 680775582 nanoseconds, so single one is around 680

@tt-vjovanovic

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,263 @@
+/*
```
Small scenario to verify:
1. Start application that uses UMD and opens this mutex.
2. Lock this mutex.
3. Start another application that uses this mutex - this should hang since it should wait for mutex.
4. Delete file in shm directory
5. Start third application that uses this mutex - this should pass since it will create new file.

If this test behaves as explained, then this mechanism might not be the best one for our use case...
On the other hand it might be good enough solution (if we say user should not do that :D)

broskoTT replied:
Since that scenario is far from somebody doing it unintentionally, I don't want to handle it. There are for sure many other ways to break our code (in whichever layer) intentionally.
The implementation relies on everyone being good, actually using the locks and not corrupt underlying files, including deleting them

broskoTT replied:
I'll add comment to the code

broskoTT replied:
Added comment

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,263 @@
+    CriticalSectionScopeGuard(int fd, pthread_mutex_t* pthread_mutex, std::string_view mutex_name) :
+        fd_(fd), pthread_mutex_(pthread_mutex), mutex_name_(mutex_name) {
+        if (flock(fd_, LOCK_EX) != 0) {
+            log_fatal("flock failed for mutex {} errno: {}", mutex_name_, std::to_string(errno));
```
This doesn't throw any more, right?

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,263 @@
+
+    ~CriticalSectionScopeGuard() noexcept {
+        // Use best effort to unlock the mutex and the flock and report warnings if something fails.
+        int err = pthread_mutex_unlock(pthread_mutex_);
```
What will happen on this line if flock fails in constructor? Can this unlock mutex for other thread?

tt-vjovanovic replied:
If `log_fatal` throws exception, then I understand how it will work :)

broskoTT replied:
yes, it throws :) And the answer is that the destructor won't be called

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,263 @@
+
+void RobustMutex::initialize() {
+    int err;
+    bool created = open_shm_file(mutex_name_);
```
`created` variable doesn't seem to be used

broskoTT replied:
yeah left it from the first version, will remove

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,263 @@
+}
+
+void RobustMutex::lock() {
+    int lock_res = pthread_mutex_lock(&(mutex_wrapper_ptr_->mutex));
```
Can we verify simple scenario:
Start two applications. Open same mutex of this type. Ask for lock.
Process A will get the lock, while process B will wait for process A to unlock it.
Intentionally crash process A (with terminate or unhandled exception).
Verify that process B entered `lock_res == EOWNERDEAD` and it got the lock.

broskoTT replied:
I did similar scenario manually, I commented on the way I did manual testing in my very long PR description. I ran one process and crashed it while holding the lock. Then ran it again and observed some added prints which showed this code path being executed.
However I didn't test if the other is waiting on a lock at the moment when the first one crashes....

broskoTT replied:
Yeah just did this test as well

```
RobustMutex mutex("test_mutex");
mutex.initialize();
mutex.lock();
std::cout << "locked mutex" << std::endl;
while(true);
```

Also added prints to this codepath.
Ran it in two windows, then when I CTRL+C in the first, the second gets the lock.

joelsmithTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,263 @@
+    return created;
+}
+
+void RobustMutex::open_pthread_mutex(int shm_fd_) {
```
Is the `int shm_fd_` parameter needed?

broskoTT replied:
I started with another kind of implementation, so this is leftover. Will change

broskoTT commented on device/api/umd/device/utils/robust_mutex.h:
```diff
@@ -0,0 +1,70 @@
+    // The initialization can fail and throw an exception. In case of failure we still want to clean up the resources.
+    // For easier handling of such case, the destructor cleans up the resources if they were taken. Having this code out
+    // of constructor will guarantee that the destructor is called in case of exception.
+    void initialize();
```
This was @tt-vjovanovic 's suggestion. Doing it separately allows me to always clean up in the destructor if the initialization fails, instead of trying to mimick some other way to implement try-catch-finally block in c++, or try to release resources on each path when possible.
You can actually see that in the constructor of CriticalSectionScopeGuard, but there is only one place where it can fail which requires cleanup, so it's not that bad. When I tried to do the same in the initialize function it got messy

joelsmithTT replied:
I can see how it gets messy quickly.

broskoTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,263 @@
+    CriticalSectionScopeGuard(int fd, pthread_mutex_t* pthread_mutex, std::string_view mutex_name) :
+        fd_(fd), pthread_mutex_(pthread_mutex), mutex_name_(mutex_name) {
+        if (flock(fd_, LOCK_EX) != 0) {
+            log_fatal("flock failed for mutex {} errno: {}", mutex_name_, std::to_string(errno));
```
But this is just the naming issue right? We can change log_fatal to log_exception or something in another PR, which would make it clearer

joelsmithTT replied:
yes / ok

joelsmithTT commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,263 @@
+
+    size_t file_size = get_file_size(shm_fd_);
+    // This means the file was just created but nor file nor mutex were initialized.
+    bool should_initialize = (file_size == 0);
```
The problem to solve is the case where `ftruncate()` has sized the file appropriately but the process dies before `initialize_pthread_mutex_first_use()` succeeds.  You do not need to worry about multiple simultaneous processes because the flock technique serializes access here.

As written, the following code path can get skipped even if there was never a successful call to `initialize_pthread_mutex_first_use()`:

```
    if (should_initialize) {
        // We need to initialize the mutex here, since it is the first time it is being used.
        initialize_pthread_mutex_first_use();
    }
```
because `should_initialize` is based only on the size of the segment, not on whether the mutex in the segment was actually initialized.

joelsmithTT replied:
Yes, I believe automatically recovering from any partial-initialization crash scenario is preferable to requiring manual intervention and aligns better with the 'Robust' goal.

The key change is in how the decision to call `initialize_pthread_mutex_first_use()` is made inside the `CriticalSectionScopeGuard` block.  Currently, that decision hinges on whether the file size is zero.

Instead, after mapping the memory, the logic can check `if (mutex_wrapper_ptr_->initialized != INITIALIZED_FLAG)` and only call `initialize_pthread_mutex_first_use()` if that condition is true.  I think this will work even if a previous process was killed while inside of the `pthread_mutex_init()`.

Do you see any holes in this logic?

broskoTT replied:
Yes, you have to zero out the mutex memory before you try to initialize it again, but that's also doable.
Why do I fail then in case of file size being different than the expected? in that case I should also then try to resize it to proper size?

broskoTT replied:
I changed the code, so it will try to initialize now in case of wrong file size, or wrong INITIALIZED_FLAG, and will report warnings if something wasn't as expected.

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,295 @@
+static constexpr int ALL_RW_PERMISSION = 0666;
+static constexpr std::string_view UMD_LOCK_PREFIX = "TT_UMD_LOCK.";
+// Any value which is unlikely to be found at random in the memory.
+static constexpr uint64_t INITIALIZED_FLAG = 0x5454554d444d5458;  // TTUMDMTX
```
Stupid question, but easy to try :)
There was character literal, might be Microsoft compiler extension, but you were able to write something like `'TEST'`, which will generate 32bit integer with character values in reverse order...

tt-vjovanovic replied:
ChatGPT says it is Microsoft extension :)

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,295 @@
+}
+
+void RobustMutex::initialize() {
+    int err;
```
`err` doesn't seem to be used

tt-vjovanovic commented on device/utils/robust_mutex.cpp:
```diff
@@ -0,0 +1,295 @@
+    bool file_was_resized = resize_shm_file();
+
+    // We now open the mutex in the shared memory file.
+    open_pthread_mutex();
```
Should this be done after file initialization (call to `initialize_pthread_mutex_first_use`)?

broskoTT commented on .github/workflows/run-tests.yml:
```diff
@@ -78,9 +78,6 @@ jobs:
         run: |
           ${{ env.TEST_OUTPUT_DIR }}/umd/misc/umd_misc_tests
 
-      # TODO: Re-enable unified tests in CI once we solve the bugs
```
Verify tt-metal tests pass
All runs on brosko/test_umdlocks :
- [ ] All post-commit tests : https://github.com/tenstorrent/tt-metal/actions/runs/14335808671
- [ ] Blackhole post-commit tests : https://github.com/tenstorrent/tt-metal/actions/runs/14335810810
- [ ] (Single-card) Model perf tests : https://github.com/tenstorrent/tt-metal/actions/runs/14335813192
- [ ] (Single-card) Device perf regressions : https://github.com/tenstorrent/tt-metal/actions/runs/14335816204
- [ ] (T3K) T3000 unit tests : https://github.com/tenstorrent/tt-metal/actions/runs/14335819889
- [ ] (T3K) T3000 demo tests : https://github.com/tenstorrent/tt-metal/actions/runs/14335822736
- [ ] (TG) TG unit tests : https://github.com/tenstorrent/tt-metal/actions/runs/14335825045
- [ ] (TG) TG demo tests : https://github.com/tenstorrent/tt-metal/actions/runs/14335826785
